### PR TITLE
add ExponentialBackoffStrategy with exponential_retry helper

### PR DESCRIPTION
**Pull request by DavydenkovM**
PR related to the following issue: https://github.com/gocardless/activejob-retry/issues/31

### CODE REVIEW

File: lib/active_job/retry/exponential_backoff_strategy.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    class ExponentialBackoffStrategy < ConstantBackoffStrategy
+      def initialize(options)
+        ExponentialOptionsValidator.new(options).validate!
+        @retry_limit          = options.fetch(:limit, 1)
+        @fatal_exceptions     = options.fetch(:fatal_exceptions, [])
+        @retryable_exceptions = options.fetch(:retryable_exceptions, nil)
+      end
+
+      def retry_delay(attempt, _exception)
+        (attempt**4 + 15 + (rand(30) * (attempt + 1))).seconds
```
Comment (isaacseymour) [inline lib/active_job/retry/exponential_backoff_strategy.rb]: I'd prefer some space around the `**` operator here

File: lib/active_job/retry/exponential_options_validator.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  module Retry
+    class ExponentialOptionsValidator
+      def initialize(options)
+        @options = options
+      end
+
+      def validate!
+        validate_limit_numericality!
+        validate_infinite_limit!
+        validate_delay_not_specified!
```
Comment (DavydenkovM) [inline lib/active_job/retry/exponential_options_validator.rb]: This is the only difference with respect to the `ConstantOptionsValidator`

File: spec/retry/exponential_backoff_strategy_spec.rb [hunk 3]
```diff
@@ -0,0 +1,145 @@
+require 'spec_helper'
+
+RSpec.describe ActiveJob::Retry::ExponentialBackoffStrategy do
+  let(:strategy) { described_class.new(options) }
+
+  describe '#should_retry?' do
```
Comment (DavydenkovM) [inline spec/retry/exponential_backoff_strategy_spec.rb]: similar to the `ConstantBackoffStrategy`, but I provide additional tests for `#retry_delay` below

File: spec/retry/exponential_options_validator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      let(:options) { { limit: nil } }
+      it { is_expected.to raise_error(ActiveJob::Retry::InvalidConfigurationError) }
+    end
+
+    context 'accidental finite limit' do
+      let(:options) { { unlimited_retries: true } }
+      it { is_expected.to raise_error(ActiveJob::Retry::InvalidConfigurationError) }
+    end
+
+    context 'delay provided' do
```
Comment (DavydenkovM) [inline spec/retry/exponential_options_validator_spec.rb]: if delay provided - raise `InvalidConfigurationError`

File: spec/retry/exponential_backoff_strategy_spec.rb [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        expect(subject).to be < attempt_3_delay
+        expect(attempt_3_delay).to be < attempt_5_delay
+      end
+    end
+
+    context 'unlimited retries' do
+      let(:options) { { limit: nil, unlimited_retries: true } }
+      let(:attempt) { 1000 }
+
+      it 'not rises an error if returned delay too big' do
```
Comment (isaacseymour) [inline spec/retry/exponential_backoff_strategy_spec.rb]: I think this one could just be `specify { expect { subject }.to_not raise_error }`

File: lib/active_job/retry/exponential_backoff_strategy.rb (hunk 1 above)
Comment (DavydenkovM) [inline lib/active_job/retry/exponential_backoff_strategy.rb]: Ok, I'll fix it. (in bbatsov style guide there is no space for ** operator https://github.com/bbatsov/ruby-style-guide)

File: spec/retry/exponential_backoff_strategy_spec.rb (hunk 5 above)
Comment (DavydenkovM) [inline spec/retry/exponential_backoff_strategy_spec.rb]: thanks

**Comment by isaacseymour** [comment]
One teeny rubocop failure to fix and then I'm :+1:

File: lib/active_job/retry/exponential_backoff_strategy.rb (hunk 1 above)
Comment (DavydenkovM) [inline lib/active_job/retry/exponential_backoff_strategy.rb]: If no spaces then rubocop fails

**Comment by DavydenkovM** [comment]
done

**Comment by isaacseymour** [comment]
This looks great, thanks! Will get it pushed to RubyGems.

**Comment by isaacseymour** [comment]
Pushed as v0.5.0. Thanks so much for this!

**Pull request merged**